### PR TITLE
Decode curve gauge transactions

### LINKED ISSUE #5676: Support curve gauges
## Abstract

Currently rotki does not decode curve gauges. It handles the deposit as "Send".

Example curve gauge:
[Curve.fi: sUSDv2 Liquidity Gauge](https://etherscan.io/address/0xa90996896660decc6e997655e065b23788857849)

## Motivation

Improve features and user friendliness of rotki.

## Specification

Take care of curve gauges to decode them correctly.

- [x] Periodically query and store in general cache all curve gauges.
- [x] Decode deposit in gauge events (note: deposits can be done on behalf of a different address than the caller)
- [x] Decode withdraw from gauge events
- [x] Decode claim rewards events (note: rewards can be claimed to a different address than the caller)

### PR DESCRIPTION
**Pull request by nebolax**
Closes #5676

## Checklist

- [ ] The PR modified the frontend, and updated the [user guide](https://github.com/rotki/rotki/blob/develop/docs/usage_guide.rst) to reflect the changes.

### CODE REVIEW

File: rotkehlchen/chain/ethereum/modules/curve/curve_cache.py [hunk 1]
```diff
@@ -162,11 +169,12 @@ def ensure_curve_tokens_existence(
...
 
 
-def save_curve_pools_to_cache(
+def save_curve_data_to_cache(
         write_cursor: DBCursor,
         pools_mapping: CURVE_POOLS_MAPPING_TYPE,
+        gauges: dict[ChecksumEvmAddress, ChecksumEvmAddress],
 ) -> None:
-    """Receives data about curve pools and saves lp tokens, pool addresses and
+    """Receives data about curve pools and gauges and saves lp tokens, pool addresses, gauges and
```
Comment (prettyirrelevant) [inline rotkehlchen/chain/ethereum/modules/curve/curve_cache.py:177]:
```suggestion
    """Stores data received about curve pools and gauges in the cache.
```

File: rotkehlchen/chain/ethereum/node_inquirer.py [hunk 2]
```diff
@@ -263,11 +265,24 @@ def assure_curve_protocol_cache_is_queried(self) -> bool:
...
+            deployed_block=0,  # deployment_block is not used and the contract is dynamic
+        )
+        pools = query_curve_registry_pools(ethereum=self, registry_contract=registry_contract)  # noqa: E501
         pools.update(query_curve_meta_pools(ethereum=self, curve_address_provider=curve_address_provider))  # noqa: E501
         updated_pools = ensure_curve_tokens_existence(ethereum_inquirer=self, pools_mapping=pools)
+        pools_addresses = [pool_data[0] for pool_data in pools.values()]
+        gauges = query_curve_gauges(ethereum=self, registry_contract=registry_contract, known_pools=pools_addresses)  # noqa: E501
         with GlobalDBHandler().conn.write_ctx() as write_cursor:
-            save_curve_pools_to_cache(write_cursor=write_cursor, pools_mapping=updated_pools)
+            save_curve_data_to_cache(write_cursor=write_cursor, pools_mapping=updated_pools, gauges=gauges)  # noqa: E501
```
Comment (prettyirrelevant) [inline rotkehlchen/chain/ethereum/node_inquirer.py:285]: you can break this into multiple lines

**Review by prettyirrelevant — COMMENTED** [review]
the only additions to the global db are the gauges right?

**Comment by nebolax** [comment]
> the only additions to the global db are the gauges right?

Yes

File: rotkehlchen/chain/ethereum/modules/curve/accountant.py [hunk 3]
```diff
@@ -35,4 +35,22 @@ def event_settings(self, pot: 'AccountingPot') -> dict[str, TxEventSettings]:  #
...
+            get_tx_event_type_identifier(HistoryEventType.WITHDRAWAL, HistoryEventSubType.REMOVE_ASSET, CPT_CURVE): TxEventSettings(  # noqa: E501
+                taxable=False,
+                count_entire_amount_spend=False,
+                count_cost_basis_pnl=False,
+                method='acquisition',
+            ),
+            get_tx_event_type_identifier(HistoryEventType.RECEIVE, HistoryEventSubType.REWARD, CPT_CURVE): TxEventSettings(  # noqa: E501
+                taxable=True,
+                count_entire_amount_spend=False,
+                count_cost_basis_pnl=False,
```
[the remark this answers is not in the document]
Comment (nebolax) [reply rotkehlchen/chain/ethereum/modules/curve/accountant.py:53]: Will change, but for acquisitions this doesn't matter anyway https://github.com/rotki/rotki/blob/cf06565348288656b3dc02f75e5b05b030454e5a/rotkehlchen/accounting/pot.py#L193

File: rotkehlchen/chain/ethereum/modules/curve/decoder.py [hunk 4]
```diff
@@ -258,8 +262,40 @@ def _decode_curve_events(
...
+                crypto_asset = event.asset.resolve_to_crypto_asset()
+                if (
+                    event.location_label == provider and
+                    event.address == gauge_address and
+                    event.balance.amount == asset_normalized_value(amount=raw_amount, asset=crypto_asset)  # noqa: E501
+                ):
+                    if tx_log.topics[0] == GAUGE_DEPOSIT:
+                        event.event_type = HistoryEventType.DEPOSIT
+                        event.event_subtype = HistoryEventSubType.DEPOSIT_ASSET
+                        event.notes = f'Deposit {event.balance.amount} {crypto_asset.symbol} into curve gauge'  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/curve/decoder.py:287]: Add more information to know which gauge name or address it is. Name is better if you can find it, otherwise address

```suggestion
                        event.notes = f'Deposit {event.balance.amount} {crypto_asset.symbol} into {gauge_name_or_address} curve gauge'  # noqa: E501
```

File: rotkehlchen/chain/ethereum/modules/curve/decoder.py [hunk 5]
```diff
@@ -258,8 +262,40 @@ def _decode_curve_events(
...
+                        event.counterparty = CPT_CURVE
+                    else:  # Withdraw
+                        event.event_type = HistoryEventType.WITHDRAWAL
+                        event.event_subtype = HistoryEventSubType.REMOVE_ASSET
+                        event.notes = f'Deposit {event.balance.amount} {crypto_asset.symbol} from curve gauge'  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/curve/decoder.py:292]: copy pasta. Also same comment as above

```suggestion
                        event.notes = f'Withdraw {event.balance.amount} {crypto_asset.symbol} from curve gauge'  # noqa: E501
```

File: rotkehlchen/chain/ethereum/modules/curve/decoder.py [hunk 6]
```diff
@@ -286,15 +322,29 @@ def _maybe_enrich_curve_transfers(
...
             event.notes = f'Receive {event.balance.amount} {crypto_asset.symbol} from the curve pool {CURVE_Y_DEPOSIT}'  # noqa: E501
             return True
+        if (
+            event.event_type == HistoryEventType.RECEIVE and
+            event.event_subtype == HistoryEventSubType.NONE and
+            source_address in self.curve_gauges
+        ):
+            crypto_asset = event.asset.resolve_to_crypto_asset()
+            event.event_subtype = HistoryEventSubType.REWARD
+            event.notes = f'Receive {event.balance.amount} {crypto_asset.symbol} rewards from curve'  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/curve/decoder.py:332]: Add the curve gauge or name here

```suggestion
            event.notes = f'Receive {event.balance.amount} {crypto_asset.symbol} rewards from curve'  # noqa: E501
```

File: rotkehlchen/chain/ethereum/node_inquirer.py [hunk 7]
```diff
@@ -263,11 +265,28 @@ def assure_curve_protocol_cache_is_queried(self) -> bool:
             return False
 
         curve_address_provider = self.contracts.contract('CURVE_ADDRESS_PROVIDER')
-        pools = query_curve_registry_pools(ethereum=self, curve_address_provider=curve_address_provider)  # noqa: E501
+        curve_address_provider = self.contracts.contract('CURVE_ADDRESS_PROVIDER')
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/node_inquirer.py:268]: what are you doing here? You are reassigning the same variable by duplicating the line
```suggestion
```

File: rotkehlchen/data/global.db
Comment (LefterisJP) [inline rotkehlchen/data/global.db:1]: I think Isaac already asked it but just to make sure. This just contains the added curve gauges as of your PR, right?

**Review by LefterisJP — COMMENTED** [review]
So left some questions/comments.

The biggest questions I have are:

1. Regardning the accounting and how it's handle for deposits and gauge staking and how it's not tested at all at the moment.
2. Something I did not mention in the comments yet, but it's concerning to me. I really find this kind of caching we do ugly. We have some issues to improve this and they should be worked on at some point:
    a. global db: https://github.com/rotki/rotki/issues/5631
    b. user db: https://github.com/rotki/rotki/issues/5684

File: docs/changelog.rst [hunk 8]
```diff
@@ -3,6 +3,7 @@ Changelog
 =========
 
 * :feature:`5798` Users will now be able to see the avatar of the ENS.
+* :feature:`5676` Curve gauge deposit, withdrawal and claim rewards actions will now be correctly decoded.
```
[the remark this answers is not in the document]
Comment (nebolax) [reply docs/changelog.rst:6]: The only mappings used in gauges decoder are `deposit` + `deposit asset` and `withdrawal` + `remove asset`. These combinations are widely used in other decoders too, so I guess frontend already supports them. But I will ask @lukicenturi to confirm.

File: rotkehlchen/chain/ethereum/modules/curve/curve_cache.py [hunk 9]
```diff
@@ -345,3 +354,22 @@ def query_curve_meta_pools(
...
+        ethereum: 'EthereumInquirer',
+        registry_contract: 'EvmContract',
+        known_pools: list[ChecksumEvmAddress],
+) -> dict[ChecksumEvmAddress, ChecksumEvmAddress]:
+    """Queries curve gauges and returns a mapping pool address -> gauge address"""
+    pools_to_gauges: dict[ChecksumEvmAddress, ChecksumEvmAddress] = {}
+    for pool_address in known_pools:
+        gauge_address = registry_contract.call(
+            node_inquirer=ethereum,
+            method_name='get_gauges',
```
[the remark this answers is not in the document]
Comment (nebolax) [reply rotkehlchen/chain/ethereum/modules/curve/curve_cache.py:369]: There is a little confusion. There is always only one gauge per pool https://curve.readthedocs.io/dao-gauges.html#liquidity-gauges
image
 
However this particular contract method can return multiple addresses. I am not sure why it is designed like that, but with samples that I tested with the only returned address was the first address, all other were empty.

File: rotkehlchen/chain/ethereum/modules/curve/decoder.py (hunk 4 above)
Comment (nebolax) [reply rotkehlchen/chain/ethereum/modules/curve/decoder.py:287]: Hmm no, probably can't know gauge name without querying the chain

Comment (LefterisJP) [reply rotkehlchen/chain/ethereum/modules/curve/decoder.py:287]: Can't it be saved as part of the cache?

File: rotkehlchen/data/global.db
Comment (nebolax) [reply rotkehlchen/data/global.db:1]: Yes. Only gauges.

File: rotkehlchen/chain/ethereum/modules/curve/decoder.py (hunk 4 above)
Comment (nebolax) [reply rotkehlchen/chain/ethereum/modules/curve/decoder.py:287]: Technically is of course possible, but we will probably need either:
- A reverse mapping `gauge address` -> `pool address` and use pools coins symbols as the gauge name.
- Or a reverse mapping `gauge address` -> `pool address` -> `lp token` and use lp token name/symbol.

Of course it is doable, but is it worth it?

Comment (LefterisJP) [reply rotkehlchen/chain/ethereum/modules/curve/decoder.py:287]: Ah so there is no name for the gauge? Can't you just have a pool name then instead? Since this is needed anyway? And then by mapping pool to gauge you also have the pool name.

And is it worth it? Doesn't sound hard to me. And the gain is big yes. If the user has to click to etherscan to better understand what's happening we have lost.

Comment (nebolax) [reply rotkehlchen/chain/ethereum/modules/curve/decoder.py:287]: Alright. Pools also don't have names btw :), but I can make names generated from symbols of pool tokens. Does it sound good to you?

Comment (LefterisJP) [reply rotkehlchen/chain/ethereum/modules/curve/decoder.py:287]: For this PR let's just add the address of the gauge. For pools you also use the address I think, right? And then maybe let's make an issue/note for this, since it should not be limited to curve.

Comment (nebolax) [reply rotkehlchen/chain/ethereum/modules/curve/decoder.py:287]: > For pools you also use the address I think, right?

Yes.

Made an issue: #5809

File: rotkehlchen/chain/ethereum/modules/curve/curve_cache.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (LefterisJP) [reply rotkehlchen/chain/ethereum/modules/curve/curve_cache.py:369]: Judging the world by 1-2 tests you make is a guarantee to write wrong code and introduce serious bugs that will take many working days to solve. Please don't ever conclude "I don't know why they do that, but I will just assume it only ever returns a single gauge".

If you want to do that then prove to me and understand why that is. Which means studyingthe code and asking in their discord/telegram whatever they have.

If you read the [code](https://etherscan.io/address/0x90E00ACe148ca3b23Ac1bC8C240C2a7Dd9c2d7f5#code).

You immediately see it's multiple gauges:

```vyper
@view
@external
def get_gauges(_pool: address) -> (address[10], int128[10]):
    """
    @notice Get a list of LiquidityGauge contracts associated with a pool
    @param _pool Pool address
    @return address[10] of gauge addresses, int128[10] of gauge types
    """
    liquidity_gauges: address[10] = empty(address[10])
    gauge_types: int128[10] = empty(int128[10])
    gauge_controller: address = self.gauge_controller
    for i in range(10):
        gauge: address = self.liquidity_gauges[_pool][i]
        if gauge == ZERO_ADDRESS:
            break
        liquidity_gauges[i] = gauge
        gauge_types[i] = GaugeController(gauge_controller).gauge_types(gauge)

    return liquidity_gauges, gauge_types
```


But then reading the code and the docs you see that it's a for loop, of 10 possible gauges per pool. And the reason it's  10 is because there is multiple liquidity gauge types: https://curve.readthedocs.io/dao-gauges.html#gauge-types

And while these are mostly other chains it's also a differentiation of crypto pools and stable pools in ethereum.

File: rotkehlchen/chain/ethereum/node_inquirer.py [hunk 10]
```diff
@@ -263,11 +265,27 @@ def assure_curve_protocol_cache_is_queried(self) -> bool:
             return False
 
         curve_address_provider = self.contracts.contract('CURVE_ADDRESS_PROVIDER')
-        pools = query_curve_registry_pools(ethereum=self, curve_address_provider=curve_address_provider)  # noqa: E501
+        get_registry_result = curve_address_provider.call(
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/node_inquirer.py:268]: Thinking of this shouldn't you also add a DB migration here for this? To make sure that curve gauges are queried? Otherwise a user who was running 1.27.1 will not see any gauges data until `should_update_protocol_cache(GeneralCacheType.CURVE_LP_TOKENS)` is `True` again.

Perhaps just resetting last update ts for all of them would do it?

Or another approach is to add a new `should_update_protocol_cache(GeneralCacheType.CURVE_GAUGES)` variable, which would then not be set and as such need no migration.

I guess updating them separately is not needed so perhaps just adding a small deletion of last_update in the migration is best.

File: rotkehlchen/chain/ethereum/modules/curve/accountant.py [hunk 11]
```diff
@@ -35,4 +35,22 @@ def event_settings(self, pot: 'AccountingPot') -> dict[str, TxEventSettings]:  #
                 method='spend',
                 accountant_cb=self._process_deposit_or_withdrawal,
             ),
+            get_tx_event_type_identifier(HistoryEventType.DEPOSIT, HistoryEventSubType.DEPOSIT_ASSET, CPT_CURVE): TxEventSettings(  # noqa: E501
+                taxable=False,
+                count_entire_amount_spend=False,
+                count_cost_basis_pnl=False,
+                method='spend',
+            ),
+            get_tx_event_type_identifier(HistoryEventType.WITHDRAWAL, HistoryEventSubType.REMOVE_ASSET, CPT_CURVE): TxEventSettings(  # noqa: E501
```
[the remark this answers is not in the document]
Comment (LefterisJP) [reply rotkehlchen/chain/ethereum/modules/curve/accountant.py:44]: Yes it does make sense to move to another PR, so long as when you hear me every day saying I need all these solved properly you understand that this means you guys need to fix those and not just move onto other issues and forget about it :wink:

**Review by LefterisJP — COMMENTED** [review]
Left 1 definite todo item and then also the multiple gauges per pool problem.

If there is a possiility of multiple gauges per pool we need to handle it and we need to do it from the get-go, not later. Since we built a cache of pools to gauge, if we build it wrong from the start then we are fucked. We need the cache to be correct since it persists.

I asked some questions to curve myself too, to see what they explain.

File: rotkehlchen/chain/ethereum/modules/curve/curve_cache.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (nebolax) [reply rotkehlchen/chain/ethereum/modules/curve/curve_cache.py:369]: I did a more thorough check and checked this method against all our known curve pools. There were different types of gauges returned, however there was always a single gauge (or no gauges) returned. Also if we consider the possibility that there are multiple gauges returned, how do you think we should handle it?

File: rotkehlchen/globaldb/migrations/migration1.py [hunk 12]
```diff
@@ -62,12 +72,27 @@ def globaldb_data_migration_1(conn: 'DBConnection') -> None:
         # Remove old setting if existing
         write_cursor.execute('DELETE FROM settings WHERE name="last_assets_json_version";')
 
-        # Add a makerdao vault types cache, at a time that will allow refresh
-        timestamp = Timestamp(ts_now() - ETH_PROTOCOLS_CACHE_REFRESH - 1)
+        # Add a makerdao vault types cache and curve pools cache, at a time that will allow refresh
+        refreshing_timestamp = Timestamp(ts_now() - ETH_PROTOCOLS_CACHE_REFRESH - 1)
```
Comment (LefterisJP) [inline rotkehlchen/globaldb/migrations/migration1.py:76]: Hmm what are you doing? You are editing a migration that has already ran ...

File: rotkehlchen/tests/data/v4_global_before_migration1.db
Comment (LefterisJP) [inline rotkehlchen/tests/data/v4_global_before_migration1.db:1]: Undo this please

File: rotkehlchen/tests/unit/globaldb/test_migrations.py [hunk 13]
```diff
@@ -2,18 +2,24 @@
...
 from rotkehlchen.globaldb.migrations.migration1 import ILK_REGISTRY_ABI, ilk_mapping
 from rotkehlchen.tests.utils.globaldb import patch_for_globaldb_migrations
+from rotkehlchen.types import GeneralCacheType
+from rotkehlchen.utils.misc import ts_now
 
 
 @pytest.mark.parametrize('globaldb_upgrades', [[]])
 @pytest.mark.parametrize('run_globaldb_migrations', [False])
 @pytest.mark.parametrize('custom_globaldb', ['v4_global_before_migration1.db'])
+@freeze_time('2023-03-20')  # freezing time just to make sure comparisons of timestamps won't fail
```
Comment (LefterisJP) [inline rotkehlchen/tests/unit/globaldb/test_migrations.py:22]: undo this

File: rotkehlchen/chain/ethereum/modules/curve/curve_cache.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (LefterisJP) [reply rotkehlchen/chain/ethereum/modules/curve/curve_cache.py:369]: We handle it by keeping a mapping of one to many if it's possible. The point is to find if it is possible by asking their engineers. As I said ... manual experimenting is pointless here. Can't sample your way into what is possible. You need to ask and get more information.

If multiple pool gauges are possible then we should keep a mapping of pool -> list of gauges, and take that into account in decoding. Doesn't that make sense to you?

File: rotkehlchen/globaldb/cache.py [hunk 14]
```diff
@@ -69,7 +69,7 @@ def globaldb_get_general_cache_values_like(
 ) -> list[str]:
     """Function to read globaldb cache. Returns all the values that start with key."""
     cache_key = _compute_cache_key(key_parts)
-    cursor.execute('SELECT value FROM general_cache WHERE key=?', (cache_key,))
+    cursor.execute('SELECT value FROM general_cache WHERE key LIKE ?', (f'{cache_key}%',))
```
Comment (LefterisJP) [inline rotkehlchen/globaldb/cache.py:72]: What are you doing here? Why this change? Reading the name and the docstring it sounds correct, but why was it equality before?

**Review by LefterisJP — COMMENTED** [review]
Still not ready. Not close. Also will need rebase on develop.

File: rotkehlchen/globaldb/cache.py (hunk 14 above)
Comment (nebolax) [reply rotkehlchen/globaldb/cache.py:72]: I suppose it was an error in the implementation. Maybe copy-paste. This function is not used in the code atm, so it was not affecting anything

Comment (LefterisJP) [reply rotkehlchen/globaldb/cache.py:72]: If it's not used at all whatsoever then remove it. It's a pleasure to remove un-needed things. If we nedit we add something like that.

File: rotkehlchen/globaldb/migrations/migration2.py [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
+
+
+def globaldb_data_migration_2(conn: 'DBConnection') -> None:
+    """Introduced at 1.28.0
+    - Resets curve cache last queried timestamp to query and save gauges
+    """
+    with conn.write_ctx() as write_cursor:
```
Comment (LefterisJP) [inline rotkehlchen/globaldb/migrations/migration2.py:24]: My bad @nebolax. I mentioned migration, but it's not needed. There is already a DB upgrade scheduled for global DB 1.28. Unless I am mistaken v4_v5 global DB upgrade comes at 1.28. Which means we do not need a migration if there is already a DB ugprade. Just add this code as part of the global DB upgrade and test it there.

No need to have a DB and a migration in the same release (unless migration was a long running process).

File: rotkehlchen/tests/unit/globaldb/test_migrations.py [hunk 16]
```diff
@@ -50,3 +55,32 @@ def test_migration1(globaldb):
                 (f'MAKERDAO_VAULT_ILK{ilk}',),
             )
             assert json.loads(cursor.fetchone()[0]) == list(info)
+
+
+@pytest.mark.parametrize('globaldb_upgrades', [[]])
+@pytest.mark.parametrize('run_globaldb_migrations', [False])
+@pytest.mark.parametrize('custom_globaldb', ['v5_global_before_migration2.db'])
+@freeze_time('2023-03-20')  # freezing time just to make sure comparisons of timestamps won't fail
```
Comment (LefterisJP) [inline rotkehlchen/tests/unit/globaldb/test_migrations.py:63]: would need to switch the test to be part of the DB upgrade

File: rotkehlchen/tests/data/v5_global_before_migration2.db
Comment (LefterisJP) [inline rotkehlchen/tests/data/v5_global_before_migration2.db:1]: this would also need to be removed.

File: rotkehlchen/chain/ethereum/modules/curve/curve_cache.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (LefterisJP) [reply rotkehlchen/chain/ethereum/modules/curve/curve_cache.py:369]: For this let's keep it like this for now but let's add more information in a subsequent PR. I will add it in the issue.

**Review by LefterisJP — COMMENTED** [review]
1. Still need to switch the migration to be part of the existing upgrade
2. As for the multiple gauges we can leave it as is for now, though there is things we need to do to double check in the future as I will explain in the issue.

**Review by LefterisJP — APPROVED** [review]
LGTM
